File: src/main/java/org/radarcns/util/serde/RadarSerde.java
package org.radarcns.util.serde;

import org.apache.kafka.common.serialization.Serde;
import org.apache.kafka.common.serialization.Serdes;
import org.apache.kafka.streams.kstream.Windowed;
import org.apache.kafka.streams.kstream.internals.WindowedDeserializer;
import org.apache.kafka.streams.kstream.internals.WindowedSerializer;

/**
 * It generates the jsonSerializer and jsonDeserializer for the given input class
 */
public class RadarSerde<T> {

    private JsonSerializer<T> jsonSerializer;
    private JsonDeserializer<T> jsonDeserializer;

    public RadarSerde(Class<T> type) {
        this.jsonSerializer = new JsonSerializer<>();
        this.jsonDeserializer = new JsonDeserializer<>(type);
    }

    public Serde<T> getSerde(){
        return Serdes.serdeFrom(jsonSerializer, jsonDeserializer);
    }

    public Serde<Windowed<T>> getWindowed(){
        WindowedSerializer<T> windowedSerializer = new WindowedSerializer<>(jsonSerializer);
        WindowedDeserializer<T> windowedDeserializer = new WindowedDeserializer<>(jsonDeserializer);
        return Serdes.serdeFrom(windowedSerializer,windowedDeserializer);
    }
}

File: src/main/kotlin/org/radarbase/util/serde/RadarSerde.kt
package org.radarbase.util.serde

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.PropertyAccessor
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.ObjectReader
import com.fasterxml.jackson.databind.ObjectWriter
import org.apache.kafka.common.serialization.Serde
import org.apache.kafka.common.serialization.Serdes

class RadarSerde<T>(type: Class<T>) {
    companion object {
        private val MAPPER = ObjectMapper().apply {
            setVisibility(PropertyAccessor.ALL, JsonAutoDetect.Visibility.NONE)
            setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY)
        }
        internal val GENERIC_WRITER: ObjectWriter = MAPPER.writer()
        internal val GENERIC_READER: ObjectReader = MAPPER.reader()
    }

    private val jsonSerializer = JsonSerializer<T>(type)
    private val jsonDeserializer = JsonDeserializer<T>(type)

    fun getSerde(): Serde<T> = Serdes.serdeFrom(jsonSerializer, jsonDeserializer)
}
